app/src/main/java/com/github/gotify/service/GotifyMessengerService.kt

```diff
@@ -0,0 +1,176 @@
+package com.github.gotify.service
+
+import android.app.Service
+import android.content.Intent
+import android.os.*
+import androidx.annotation.RequiresApi
+import androidx.core.os.bundleOf
+import com.github.gotify.Settings
+import com.github.gotify.api.ClientFactory
+import com.github.gotify.client.api.ApplicationApi
+import com.github.gotify.log.Log
+import java.io.IOException
+
+/**
+ * THIS SERVICE IS USED BY OTHER APPS TO REGISTER
+ */
+// TODO : in the app, implement forceUnregisterApp
+
+@RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+class GotifyMessengerService : Service() {
+    /** Keeps track of all current registered clients.  */
+    private val db = MessagingDatabase(this)
+    private lateinit var settings: Settings
+
+    /**
+     * Handler of incoming messages from clients.
+     */
+    internal inner class gHandler : Handler() {
+
+        override fun handleMessage(msg: Message) {
+            when (msg.what) {
+                MSG_START -> simpleAnswer(msg, MSG_START)
+                MSG_REGISTER_CLIENT -> {
+                    val uid = msg.sendingUid
+                    val msgData = msg.data
+                    val thread = object : Thread() {
+                        override fun run() {
+                            registerApp(msgData, uid)
+                        }
+                    }
+                    thread.start()
+                    // We need to wait until the registration is done to send url & token :
+                    thread.join()
+                    sendInfo(msg)
+                }
+                MSG_UNREGISTER_CLIENT -> {
+                    val thread = object : Thread() {
+                        val uid = msg.sendingUid
+                        val msgData = msg.data
+                        override fun run() {
+                            unregisterApp(msgData, uid)
+                        }
+                    }
+                    thread.start()
+                    simpleAnswer(msg, MSG_UNREGISTER_CLIENT)
+                }
+                else -> super.handleMessage(msg)
+            }
+        }
+
+        private fun simpleAnswer(msg: Message, what: Int) {
+            try {
+                msg.replyTo?.send(Message.obtain(null, what, 0, 0))
+            } catch (e: RemoteException) {
+            }
+        }
+
+        private fun unregisterApp(msg: Bundle, clientUid: Int) {
+            val clientPackageName = msg.getString("package").toString()
+            // we only trust unregistered demands from the uid who registered the app
+            if (db.strictIsRegistered(clientPackageName, clientUid)) {
+                Log.i("Unregistering $clientPackageName uid: $clientUid")
+                deleteApp(clientPackageName)
```

Comment:
I don't think we should ever delete an application, maybe there is some important stuff in it.

Reply:
I don't know, these apps aren't created manually but automatically. Maybe we should make it optionnal

Reply:
We can ask the user for this too, something like: "App XX unregistered, delete the application with all its messages?"